sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala

```diff
@@ -132,7 +131,8 @@ private[hive] class HiveClientImpl(
       if (ret != null) {
         // hive.metastore.warehouse.dir is determined in SharedState after the CliSessionState
         // instance constructed, we need to follow that change here.
-        warehouseDir.foreach { dir =>
+        val conf = hadoopConf.asInstanceOf[Configuration]
```

Comment:
You're reintroducing the original bug here. You cannot cast this to `Configuration`, because in the `sharesHadoopClasses = false` case, there are two different instances of that class involved, and this will fail with a `ClassCastException`.

Reply:
How to reproduce it?

Reply:
See other comment.

Reply:
Thanks! Will try to reproduce it.

Reply:
Actually I'm not sure just what I suggested will work. You need to reproduce the condition to trigger the `if` above:

```
       val ret = SessionState.get
        if (ret != null) {		        if (ret != null) {
```

And I'm not sure how to do that. I'm just sure that if you do, your code will not work.

Reply:
Over heeerre ... you could say this is a good example of why getting reviews is important!